smart-py/objkt_swap_v3_test.py

```diff
@@ -531,3 +540,137 @@ def test_pause_contract():
+        # how to turn json into this?
+        metadata = sp.bytes('0x697066733a2f2f516d61794e7a7258547a354237577747574868314459524c7869646646504676775a377a364b7443377268456468')
+    ).run(
+        sender = collector1.address,
+        valid = False
+    )
+
+    # objkt number must not have changed
+    scenario.verify(marketplaceV1.data.objkt_id == 152)
+
+    # try to mint an objkt with too many royalties
+    scenario += marketplaceV1.mint_OBJKT(
+        address = artist1.address,
+        amount = 2,
+        royalties = 300,
+        metadata = sp.bytes('0x697066733a2f2f516d61794e7a7258547a354237577747574868314459524c7869646646504676775a377a364b7443377268456468')
+    ).run(
+        sender = artist1.address,
+        valid = False
+    )
+
+    # objkt number must not have changed
+    scenario.verify(marketplaceV1.data.objkt_id == 152)
+
+    # try to mint an objkt with too few royalties
+    #
+    # TODO should fail but passes currently
+    # scenario += marketplaceV1.mint_OBJKT(
+    #     address = artist1.address,
+    #     amount = 2,
+    #     royalties = 5,
+    #     metadata = sp.bytes('0x697066733a2f2f516d61794e7a7258547a354237577747574868314459524c7869646646504676775a377a364b7443377268456468')
+    # ).run(
+    #     sender = artist1.address,
+    #     valid = False
+    # )
+
+    # objkt number must not have changed
+    scenario.verify(marketplaceV1.data.objkt_id == 152)
+
+    # try to mint without metadata and fail
+    #
+    # TODO null metadata should fail but passes currently
+    #
+    # scenario += marketplaceV1.mint_OBJKT(
+    #     address = artist1.address,
+    #     amount = 1,
+    #     royalties = 200,
+    #     metadata = sp.bytes('0x00')
+    # ).run(
+    #     sender = artist1.address,
+    #     valid = False
+    # )
+
+    # objkt number must not have changed
+    scenario.verify(marketplaceV1.data.objkt_id == 152)
+
+@sp.add_test(name="Test swap v3 failure conditions")
+def test_swap_failure_conditions():
+    # Get the test environment
+    testEnvironment = get_test_environment()
+    scenario = testEnvironment["scenario"]
+    artist1 = testEnvironment["artist1"]
+    collector1 = testEnvironment["collector1"]
+    collector2 = testEnvironment["collector2"]
+    objkt = testEnvironment["objkt"]
+    marketplaceV1 = testEnvironment["marketplaceV1"]
+    marketplaceV3 = testEnvironment["marketplaceV3"]
+
+    # Mint an OBJKT
+    scenario += marketplaceV1.mint_OBJKT(
+        address=artist1.address,
+        amount=1,
+        metadata=sp.pack("ipfs://fff"),
+        royalties=100).run(sender=artist1)
+
+    scenario.verify(marketplaceV1.data.objkt_id == 153)
+
+    # Try to Swap the OBJKT but fail because artist is not admin?
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(10000),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        valid=True,
```

Comment:
Why does this fail with:

```
Message: 'FA2_NOT_OPERATOR'
```

Is it because the v1 marketplace is the operator by default and there needs to be an intermediary step?

Reply:
The problem is that the v3 contract will transfer the user1 tokens to the contract address. To do that the v3 contract needs to have operator rights. Check some of the v3 tests to see that i get them before swapping for swaps that do not fail

Reply:
See this swap in H=N. You see the it starts adding the H=N contract as operator of the token to swap and it finishes removing that right
https://tzkt.io/opLCmitnTxMZ88RxJrDq1oFQg4pj64Rgf1PDcNno53mkyEBpdSC

Reply:
I see!  Thank you!